File: bwl.Imaging/RawFrames/RawIntFrame.vb
Imports Bwl.Imaging

Public Class RawIntFrame
    Inherits BlobContainer

    Public Data As Integer()
    Public ReadOnly Property Width As Integer
    Public ReadOnly Property Height As Integer

    Public Sub New(width As Integer, height As Integer, data As Integer())
        _Width = width
        _Height = height
        Me.Data = data
        Attributes.Add("Width", width.ToString)
        Attributes.Add("Height", height.ToString)
        Blobs.Add(New IntegerBlob With {.ID = "Scan0", .Data = data})
    End Sub

    Public Sub New(bc As BlobContainer)
        MyBase.New(bc)
        Width = CInt(Attributes("Width"))
        Height = CInt(Attributes("Height"))
        Data = Blobs(0).Data
    End Sub

    Public Shared Function FromLegacyFile(filename As String) As RawIntFrame
        Dim fs As New IO.FileStream(filename, IO.FileMode.Open, IO.FileAccess.Read)
        Dim sw As New IO.StreamReader(fs)
        Dim width = sw.ReadLine
        Dim height = sw.ReadLine
        Dim arr(width * height * 3) As Integer
        For i = 0 To arr.Length - 1
            arr(i) = sw.ReadLine()
        Next
        fs.Close()
        Dim frame As New RawIntFrame(width, height, arr)
        frame.Data = arr
        Return frame
    End Function

    Public Overloads Shared Function FromFile(filename As String) As RawIntFrame
        Dim file As New RawIntFrame(BlobContainer.FromFile(filename))
        Return file
    End Function

    Private Function GetCodecInfo(ByVal format As ImageFormat) As ImageCodecInfo
        Dim codecs As ImageCodecInfo() = ImageCodecInfo.GetImageDecoders()
        Dim codec As ImageCodecInfo
        For Each codec In codecs
            If codec.FormatID = format.Guid Then
                Return codec
            End If
        Next codec
        Return Nothing
    End Function

    Public Sub SaveToJpegPair(filenameWithoutExthension As String, Optional quality As Integer = 95)
        Dim _encoderParameters As New EncoderParameters(1)
        Dim _codecInfo As ImageCodecInfo = GetCodecInfo(ImageFormat.Jpeg)
        _encoderParameters.Param(0) = New EncoderParameter(System.Drawing.Imaging.Encoder.Quality, quality)
        Dim mtrs = ConvertTo8BitPair(Me)
        Dim bmph = mtrs(0).ToBitmap
        Dim bmpl = mtrs(1).ToBitmap
        Dim fnameh = filenameWithoutExthension + "_h.jpeg"
        Dim fnamel = filenameWithoutExthension + "_l.jpeg"
        bmph.Save(fnameh, _codecInfo, _encoderParameters)
        bmpl.Save(fnamel, _codecInfo, _encoderParameters)
    End Sub

    Public Shared Function FromJpegPair(filenameWithoutExthension As String) As RawIntFrame
        Dim fnameh = filenameWithoutExthension + "_h.jpeg"
        Dim fnamel = filenameWithoutExthension + "_l.jpeg"
        Dim bmph = New Bitmap(fnameh)
        Dim bmpl = New Bitmap(fnamel)
        Dim mtrs = {bmph.BitmapToRgbMatrix, bmpl.BitmapToRgbMatrix}
        Dim frame = ConvertFrom8BitPair(mtrs)
        Return frame
    End Function
End Class

File: Bwl.Imaging/RawFrames/Integer/RawIntFrame.cs
using System.Drawing;
using System.Drawing.Imaging;

namespace Bwl.Imaging
{

    public class RawIntFrame : BlobContainer
    {

        public int[] Data;
        public int Width { get; private set; }
        public int Height { get; private set; }

        public RawIntFrame(int width, int height, int[] data)
        {
            Width = width;
            Height = height;
            Data = data;
            Attributes.Add("Width", width.ToString());
            Attributes.Add("Height", height.ToString());
            Blobs.Add(new IntegerBlob() { ID = "Scan0", Data = data });
        }

        public RawIntFrame(int width, int height) : this(width, height, Array.Empty<int>())
        {
        }

        public RawIntFrame(BlobContainer bc) : base(bc)
        {
            Width = Convert.ToInt32(Attributes["Width"]);
            Height = Convert.ToInt32(Attributes["Height"]);
            Data = Blobs[0].Data;
        }

        public void SetDataToScan0Blob(int[] data)
        {
            var scan0Blob = Blobs.First(item => item.ID == "Scan0");
            scan0Blob.Data = data;
        }

        public void SetDataToScan0Blob()
        {
            var scan0Blob = Blobs.First(item => item.ID == "Scan0");
            scan0Blob.Data = Data;
        }

        public void GetDataFromScan0Blob()
        {
            var scan0Blob = Blobs.First(item => item.ID == "Scan0");
            Data = scan0Blob.Data;
        }

        public static RawIntFrame FromLegacyFile(string filename)
        {
            using (var fs = new System.IO.FileStream(filename, System.IO.FileMode.Open, System.IO.FileAccess.Read))
            {
                using (var sw = new System.IO.StreamReader(fs))
                {
                    int width = Convert.ToInt32(sw.ReadLine());
                    int height = Convert.ToInt32(sw.ReadLine());
                    var arr = new int[width * height * 3 + 1];
                    for (int i = 0, loopTo = arr.Length - 1; i <= loopTo; i++)
                        arr[i] = Convert.ToInt32(sw.ReadLine());
                    fs.Close();
                    var frame = new RawIntFrame(width, height, arr);
                    frame.Data = arr;
                    return frame;
                }
            }
        }

        public new static RawIntFrame FromFile(string filename)
        {
            var @file = new RawIntFrame(BlobContainer.FromFile(filename));
            return @file;
        }

        private ImageCodecInfo GetCodecInfo(ImageFormat format)
        {
            ImageCodecInfo[] codecs = ImageCodecInfo.GetImageDecoders();
            foreach (var codec in codecs)
            {
                if (codec.FormatID == format.Guid)
                {
                    return codec;
                }
            }
            return null;
        }

        public void SaveToJpegPair(string filenameWithoutExthension, int quality = 95)
        {
            var col = HSV.FromRgb(1, 1, 1);
            var _encoderParameters = new EncoderParameters(1);
            var _codecInfo = GetCodecInfo(ImageFormat.Jpeg);
            _encoderParameters.Param[0] = new EncoderParameter(Encoder.Quality, quality);
            RGBMatrix[] mtrs = RawIntFrameConverters.ConvertTo8BitPair(this);
            var bmph = mtrs[0].ToBitmap();
            var bmpl = mtrs[1].ToBitmap();
            string fnameh = filenameWithoutExthension + "_h.jpeg";
            string fnamel = filenameWithoutExthension + "_l.jpeg";
            bmph.Save(fnameh, _codecInfo, _encoderParameters);
            bmpl.Save(fnamel, _codecInfo, _encoderParameters);
        }

        public static RawIntFrame FromJpegPair(string filenameWithoutExthension)
        {
            string fnameh = filenameWithoutExthension + "_h.jpeg";
            string fnamel = filenameWithoutExthension + "_l.jpeg";
            var bmph = new Bitmap(fnameh);
            var bmpl = new Bitmap(fnamel);
            RGBMatrix[] mtrs = new[] { bmph.BitmapToRgbMatrix(), bmpl.BitmapToRgbMatrix() };
            var frame = RawIntFrameConverters.ConvertFrom8BitPair(mtrs);
            return frame;
        }
    }
}